### ISSUE TITLE
istio/pkg/viperconfig/ProcessViperConfig duplicates slice and array vars

### ISSUE THREAD

**Issue by therealmitchconnors** [istio/istio#16143]
TL;DR: Using ProcessViperConfig() results in duplication of all default slice and array flag values.

## Explanation
The viperconfig package is designed to allow users of cobra command line args to quickly add viper config sources to their executable.  Ideally, this can be done by adding one line of code per root command, which iterates over all sub-commands, binding their cobra flags to viper, and adding a PreRun hook that overwrites cobra's values with viper's.  This allows flag definitions like:
```
var server string
cobraCmd.PersistentFlags().StringVar(&server, "server", "the server's address")
```
to continue functioning, but with viper values injected at pre-run time.

This architecture works by [iterating](https://github.com/istio/pkg/blob/master/viperconfig/viperconfig.go#L52) over all cobra flags and calling Set() using the viper values at pre-run time.  This pattern is successful for all but two types of flags: Slices and Arrays.  

In the case of Slices and Arrays, the Set() command first checks to see if the value has been set previously (in our scenario it always has), and if so [appends](https://github.com/spf13/pflag/blob/master/string_slice.go#L49) to the underlying list rather than replacing the list.  This allows pflag to accommodate flags that are set multiple times in a command line, but in our case, the semantic break in Set() causes us to append the viper values, rather than overwriting.
### Option 1 - Pflag Pull Request
While changing the behavior of Set() would break pflag functionality, we could add a Clear() function to the flag interface, which resets the `changed` flag, and sets the underlying value to the zero value for its type.  While the lack of generics in go would make this change a bit verbose, it would allow us and others to deterministically overwrite flag values when appropriate.  The downside is, it involves getting a PR merged into someone else's repo, and waiting on a release.

### Option 2 - Maintain a global "list" registry.
Less elegantly, we could require users of list flags (we have 17 such flags today) to register the flag and it's value pointer in the viperconfig namespace.  Then, when ProcessConfig is called, viperconfig could overwrite the value of the pointer, rather than the flag, and throw some sort of error if any slice or array flag is not registered.  This is clunky, but has the benefit of not requiring a PR to a third party repository.

**Comment by geeknoid** [comment]
Option 1 isn't viable. There has been no substantial commit to the pflag  repo in a long time, and there are nice useful PRs that haven't been merged in years. So that repo is effectively abandoned. I'm not thrilled about diverging there and permanently keeping a fork private to Istio.

Before starting to implement option 2, would you mind adding a test  case in viperconfig_test.go that show the problem that is happening? Comment the test case out and check it in. This would make it possible to experiment with potential alternate solutions.

**Comment by howardjohn** [comment]
@therealmitchconnors what is the use impact of this for users? I don't really understand this but seems like it could be pretty high impact?

**Comment by therealmitchconnors** [comment]
@howardjohn if we merged https://github.com/istio/istio/pull/16027 before resolving this issue, Istio would be very broken, as seen from the many test failures on that PR.  Before that PR is merged, there is zero impact, because this function is never called.

**Comment by therealmitchconnors** [comment]
@geeknoid Done.  
https://github.com/istio/pkg/pull/53
